### PR TITLE
[MM-53454] Add export file settings + slash command for public link

### PR DESCRIPTION
**Pull request by JulienTant**
#### Summary
- `FileSettings` now contains a `DedicatedExportStore` option that allows server admins to specify a different file configuration for exports.
- If `DedicatedExportStore` is `true`, a new ExportFileBackend() is generated using the new config values (all stored in `FileSettings`):
```
"ExportDriverName"
"ExportDirectory"
"ExportAmazonS3AccessKeyId"
"ExportAmazonS3SecretAccessKey"
"ExportAmazonS3Bucket"
"ExportAmazonS3PathPrefix"
"ExportAmazonS3Region"
"ExportAmazonS3Endpoint"
"ExportAmazonS3SSL"
"ExportAmazonS3SignV2"
"ExportAmazonS3SSE"
"ExportAmazonS3Trace"
"ExportAmazonS3RequestTimeoutMilliseconds"
"ExportAmazonS3PresignExpiresSeconds"
```
If `DedicatedExportStore` is false, the. standard filestorage is used as before (default)
- Adds a new slashcommand `/exportlink [job-id|zip file|latest]` that helps sysadmin generate a s3 presigned URL for an export. Using `latest` will automatically pull the latest export found. This command is only available for Sysadmins AND the FF `EnableExportDirectDownload` is true AND the storage is compatible with generating link (s3) AND `DedicatedExportStore` is `true` (we don't want to allow this feature if this is a shared storage
- Adds `mmctl export generate-presigned-url [exportname]` that will generate a presigned URL directly from `mmctl`

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-53454

#### Screenshots
N/A

#### Release Note
```release-note
NONE
```

### CODE REVIEW

**Comment by JulienTant** [comment]
@Willyfrog @spirosoik feel free to pull for review anything you think is necessary for this first review!

**Review by BenCookie95 — APPROVED** [review]
Code LGTM!

File: server/cmd/mmctl/docs/mmctl_export_generate-presigned-url.rst [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+mmctl export generate-presigned-url
+-----------------------------------
+
+Generate a presigned url for an export file
+
+Synopsis
+~~~~~~~~
+
+
+Generate a presigned url for an export file
```
Comment (Willyfrog) [inline server/cmd/mmctl/docs/mmctl_export_generate-presigned-url.rst:12]:
```suggestion
Generate a presigned url for an export file. This is helpful when an export is big and might have trouble downloading from the Mattermost server.
```

**Review by Willyfrog — APPROVED** [review]
Added a minor change to help understand why this is useful, but no behaviour changes.
LGTM!

**Comment by Willyfrog** [comment]
@spirosoik @esarafianou gentle reminder to review this

File: server/channels/app/slashcommands/command_exportlink.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+	}
+
+	b := a.ExportFileBackend()
+	_, ok := b.(filestore.FileBackendWithLinkGenerator)
+	if !ok {
+		return &model.CommandResponse{ResponseType: model.CommandResponseTypeEphemeral, Text: args.T("api.command_exportlink.driver.app_error")}
+	}
+
+	file := ""
+	if message == "latest" {
```
Comment (spirosoik) [inline server/channels/app/slashcommands/command_exportlink.go:70]: could you elaborate a bit more here? is this the export latest?

**Review by spirosoik — COMMENTED** [review]
Few minor comments. @gabrieljackson can help here a bit more as he has better experience with server.

File: server/channels/app/slashcommands/command_exportlink.go (hunk 2 above)
Comment (JulienTant) [reply server/channels/app/slashcommands/command_exportlink.go:70]: Yeah. If the admin uses `/exportlink latest`, we'll try to find the latest generated export and return a link to this export

File: server/channels/app/platform/service.go [hunk 3]
```diff
@@ -247,6 +248,19 @@ func New(sc ServiceConfig, options ...Option) (*PlatformService, error) {
 		ps.filestore = backend
 	}
 
+	if ps.exportFilestore == nil {
+		ps.exportFilestore = ps.filestore
+		if *ps.Config().FileSettings.DedicatedExportStore {
+			insecure := ps.Config().ServiceSettings.EnableInsecureOutgoingConnections
```
[the remark this answers is not in the document]
Comment (JulienTant) [reply server/channels/app/platform/service.go:254]: The insecure part allows server admins to have a minio instance with a self signed certificate. This is the same code we use for the normal storage. Any reason why we would not want to allow this for the backup storage?

File: server/channels/app/export.go [hunk 4]
```diff
@@ -831,16 +833,52 @@ func (a *App) ListExports() ([]string, *model.AppError) {
...
+		return nil, model.NewAppError("GeneratePresignURLForExport", "app.eport.generate_presigned_url.config.app_error", nil, "", http.StatusInternalServerError)
+	}
+
+	b := a.ExportFileBackend()
+	backend, ok := b.(filestore.FileBackendWithLinkGenerator)
+	if !ok {
+		return nil, model.NewAppError("GeneratePresignURLForExport", "app.eport.generate_presigned_url.driver.app_error", nil, "", http.StatusInternalServerError)
+	}
+
+	p := path.Join(*a.Config().ExportSettings.Directory, "/", name)
```
Comment (jupenur) [inline server/channels/app/export.go:851]: `path.Join` adds slashes automatically.

```suggestion
	p := path.Join(*a.Config().ExportSettings.Directory, name)
```

File: server/public/model/link_metadata.go [hunk 5]
```diff
@@ -182,8 +182,8 @@ func isRoundedToNearestHour(ms int64) bool {
 	return FloorToNearestHour(ms) == ms
 }
 
-// GenerateLinkMetadataHash generates a unique hash for a given URL and timestamp for use as a database key.
-func GenerateLinkMetadataHash(url string, timestamp int64) int64 {
+// GeneratePresignedURLMetadataHash generates a unique hash for a given URL and timestamp for use as a database key.
+func GeneratePresignedURLMetadataHash(url string, timestamp int64) int64 {
```
Comment (jupenur) [inline server/public/model/link_metadata.go:186]: Why was this renamed?

File: server/channels/app/slashcommands/command_exportlink.go [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+	}
+
+	if file == "" {
+		file = message
+	}
+	if !strings.HasSuffix(file, "_export.zip") {
+		return &model.CommandResponse{ResponseType: model.CommandResponseTypeEphemeral, Text: args.T("api.command_exportlink.invalid.app_error")}
+	}
+
+	res, err := a.GeneratePresignURLForExport(file)
```
Comment (jupenur) [inline server/channels/app/slashcommands/command_exportlink.go:107]: What happens if I do `/exportlink ../../oh89z5ckgby5pmbdxtkyr7tsta/export/ebesgyk6e7rqdpy5jm5hme44gw`, where `oh89z5ckgby5pmbdxtkyr7tsta` is somebody else's cloud installation ID and `ebesgyk6e7rqdpy5jm5hme44gw` is an export on that installation?

**Review by jupenur — CHANGES_REQUESTED** [review]

File: server/public/model/link_metadata.go (hunk 5 above)
Comment (JulienTant) [reply server/public/model/link_metadata.go:186]: It was a mistake, nice catch!

File: server/channels/app/slashcommands/command_exportlink.go (hunk 6 above)
Comment (JulienTant) [reply server/channels/app/slashcommands/command_exportlink.go:107]: 👍🏻 Given that we are really just expecting a filename and not a path, I'll get the base name out of the string. the path prefix for the bucket + export directory are appended to this from the config.

**Review by jupenur — APPROVED** [review]

File: server/channels/app/platform/service.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (spirosoik) [reply server/channels/app/platform/service.go:254]: I think we should aim for the best practices right? Self-signed means we have a certificate without the support of a trusted certificate authority to guarantee the validity of the certificate details.

File: server/channels/app/slashcommands/command_exportlink.go (hunk 2 above)
Comment (spirosoik) [reply server/channels/app/slashcommands/command_exportlink.go:70]: wondering is it better to make this a `const`/enum  instead versus a text in the code?

File: server/channels/app/platform/service.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (JulienTant) [reply server/channels/app/platform/service.go:254]: I honestly don't have a strong opinion on that, I figured we would like to have feature parity between the normal filestore and the filestore dedicated to export.
If it's not the case, I don't mind `insecure` to false - trying to get this in ASAP because the deadline is coming!

[the remark this answers is not in the document]
Comment (esarafianou) [reply server/channels/app/platform/service.go:254]: I'd prefer setting `insecure` to false. Feature parity is great only when we still believe that we took the best decisions and followed the best practices when we developed the initial one. Otherwise, let's keep improving :)

[the remark this answers is not in the document]
Comment (spirosoik) [reply server/channels/app/platform/service.go:254]: Agree with @esarafianou

File: server/channels/app/platform/service.go [hunk 7]
```diff
@@ -247,6 +248,19 @@ func New(sc ServiceConfig, options ...Option) (*PlatformService, error) {
...
+			backend, err2 := filestore.NewFileBackend(filestore.NewExportFileBackendSettingsFromConfig(&ps.Config().FileSettings, license != nil && *license.Features.Compliance, insecure != nil && *insecure))
```
[the remark this answers is not in the document]
Comment (JulienTant) [reply server/channels/app/platform/service.go]: Changed! Thanks!

File: server/channels/app/platform/service.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (JulienTant) [reply server/channels/app/platform/service.go]: Changed! Thanks!

**Review by esarafianou — APPROVED** [review]

**Review by gabrieljackson — APPROVED** [review]
LGTM!

**Review by spirosoik — APPROVED** [review]

**Comment by amyblais** [comment]
@JulienTant Will this need QA review?

**Review by DHaussermann — APPROVED** [review]
Tested and passed
With alot of help from Julien I was able to get a working Docker setup to validate this. 
- Confirmed `DedicatedExportStore` is functional
- Successfully used export specific values to have the export stored in a unique storage space when the export is used
- Confirmed new slash command can then be used to retrieve the file from storage location
- New slash command accepts all 3 parameters as expected
- New slash command shows accurate time to live for the link which then becomes unavailable as expected
- Time to live on the link can be changed in the configuration as expected
- New slash command can only be run by users with sysadmin role
- New export command can also be run to see the download link from mmctl as expected

LGTM!

Thanks @JulienTant for all the help on this one.

**Comment by amyblais** [comment]
@JulienTant Can you help cherry-pick this to the ``cloud`` branch?
